src/Shared/EnvironmentUtilities.cs

```diff
@@ -2,22 +2,38 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System;
+#if NET7_0_OR_GREATER
+using System.Diagnostics.CodeAnalysis;
+#endif
 using System.Runtime.InteropServices;
+using System.Text.RegularExpressions;
 
 namespace Microsoft.Build.Shared
 {
     internal static partial class EnvironmentUtilities
     {
+#if NET7_0_OR_GREATER
+        [StringSyntax(StringSyntaxAttribute.Regex)]
+#endif
+        private const string WellKnownEnvironmentPattern = @"^(MSBUILD|COMPLUS_|DOTNET_)";
+
+#if NET7_0_OR_GREATER
+        [GeneratedRegex(WellKnownEnvironmentPattern, RegexOptions.IgnoreCase | RegexOptions.CultureInvariant | RegexOptions.ExplicitCapture)]
+        private static partial Regex WellKnownEnvironmentVariableName();
+#else
+        private static Regex WellKnownEnvironmentVariableName() => s_envPropertyPattern;
+
+        private static readonly Regex s_envPropertyPattern = new(WellKnownEnvironmentPattern, RegexOptions.Compiled | RegexOptions.CultureInvariant | RegexOptions.IgnoreCase);
+#endif
+
         public static bool Is64BitProcess => Marshal.SizeOf<IntPtr>() == 8;
 
         public static bool Is64BitOperatingSystem =>
             Environment.Is64BitOperatingSystem;
 
         public static bool IsWellKnownEnvironmentDerivedProperty(string propertyName)
         {
-            return propertyName.StartsWith("MSBUILD") ||
-                propertyName.StartsWith("COMPLUS_") ||
-                propertyName.StartsWith("DOTNET_");
+            return WellKnownEnvironmentVariableName().IsMatch(propertyName);
```

Comment:
Why did you pick a regex vs. using `StringComparison.InvariantCultureIgnoreCase`? My intuition would've been that it was fast but never actually measured that.

Reply:
`InvariantCultureIgnoreCase` or `OrdinalIgnoreCase` are indeed super fast, but we're looking for multiple strings here. Doing it with a (generated) regex means we can collapse the length checks and do it only once, and (depending on strings) potentially do other sneaky match operations.

This is very unlikely to matter here: it's a `StartsWith`, both needle and haystack are short strings, this isn't a super hot path. And since we have to keep it working for net472, there's a bunch of duplicated code, which I have ALREADY screwed up (I forgot `ExplicitCapture` for the net472 case), so there's a pretty good argument that I should go back to the first commit here.

But . . . new shiny.

Reply:
I've measured with a simple regex benchmark and the three `StartsWith` calls are much faster than even the new .NET 7+ regex.

Reply:
@rainersigwald for those benchmarks to be equivalent though shouldn't the `StartsWith` be using `StringComparison.InvariantCultureIgnoreCase`?

Reply:
Wow, this is really odd! Can you re-run your benchmark passing `StringComparison.OrdinalIgnoreCase` to `StartsWith`?

Reply:
Regex still seems faster on `net7.0` at least 


| Method                       | Mean     | Error     | StdDev    | Ratio |
|----------------------------- |---------:|----------:|----------:|------:|
| Regex                        | 2.851 us | 0.0151 us | 0.0134 us |  0.54 |
| StartsWith                   | 5.289 us | 0.0259 us | 0.0243 us |  1.00 |
| StartsWithExplicitComparison | 4.021 us | 0.0123 us | 0.0109 us |  0.76 |

Reply:
Hmm

| Method     | Runtime              | Mean        | Error     | StdDev      | Ratio | RatioSD |
|----------- |--------------------- |------------:|----------:|------------:|------:|--------:|
| Regex      | .NET 7.0             |  2,699.4 ns |  35.35 ns |    33.06 ns |  0.57 |    0.01 |
| StartsWith | .NET 7.0             |    649.8 ns |   8.38 ns |     7.84 ns |  0.14 |    0.00 |
| Regex      | .NET 8.0             |  1,832.3 ns |  24.73 ns |    23.13 ns |  0.38 |    0.01 |
| StartsWith | .NET 8.0             |    464.2 ns |   8.70 ns |     8.94 ns |  0.10 |    0.00 |
| Regex      | .NET Framework 4.7.2 | 12,025.4 ns | 987.99 ns | 2,913.12 ns |  1.54 |    0.04 |
| StartsWith | .NET Framework 4.7.2 |  4,781.2 ns |  36.74 ns |    30.68 ns |  1.00 |    0.00 |

Reply:
In this table, "exact" is wrong, I meant "culture-aware"

| Method              | Runtime              | Mean        | Error       | StdDev      | Median      | Ratio | RatioSD |
|-------------------- |--------------------- |------------:|------------:|------------:|------------:|------:|--------:|
| Regex               | .NET 7.0             |  2,750.5 ns |    34.01 ns |    31.81 ns |  2,746.6 ns |  0.36 |    0.09 |
| StartsWithOrdinal   | .NET 7.0             |    656.7 ns |    11.52 ns |    10.77 ns |    653.6 ns |  0.09 |    0.02 |
| StartsWithInvariant | .NET 7.0             |  4,123.3 ns |    65.16 ns |    60.95 ns |  4,127.0 ns |  0.54 |    0.14 |
| StartsWithExact     | .NET 7.0             |  5,252.1 ns |    60.97 ns |    57.03 ns |  5,234.3 ns |  0.69 |    0.17 |
| Regex               | .NET 8.0             |  3,784.8 ns |    75.29 ns |   107.98 ns |  3,805.7 ns |  0.47 |    0.09 |
| StartsWithOrdinal   | .NET 8.0             |    907.6 ns |    43.74 ns |   128.98 ns |    955.4 ns |  0.10 |    0.01 |
| StartsWithInvariant | .NET 8.0             |  3,741.5 ns |    21.72 ns |    20.32 ns |  3,746.6 ns |  0.49 |    0.13 |
| StartsWithExact     | .NET 8.0             |  4,913.0 ns |   311.04 ns |   917.09 ns |  5,415.8 ns |  0.54 |    0.13 |
| Regex               | .NET Framework 4.7.2 | 12,578.8 ns |   248.53 ns |   394.19 ns | 12,602.8 ns |  1.54 |    0.30 |
| StartsWithOrdinal   | .NET Framework 4.7.2 |  9,265.1 ns |   419.67 ns | 1,237.40 ns |  9,202.5 ns |  1.00 |    0.00 |
| StartsWithInvariant | .NET Framework 4.7.2 | 22,333.3 ns | 1,742.88 ns | 5,138.93 ns | 24,013.8 ns |  2.51 |    0.87 |
| StartsWithExact     | .NET Framework 4.7.2 | 16,569.4 ns |   243.07 ns |   227.37 ns | 16,597.8 ns |  2.19 |    0.54 |

Reply:
This matches my benchmark where I used one fixed string for the needle. `StringComparison.OrdinalIgnoreCase` was faster than regex, even though it [regressed a bit in .NET 8](https://github.com/dotnet/runtime/issues/94713).

Reply:
I added "Handrolled" which is a lightly modified copy-paste of the .NET 8 regex generator's `TryMatchAtCurrentPosition` and which blows everything away

| Method                    | Runtime              | Mean        | Error     | StdDev    | Ratio | RatioSD |
|-------------------------- |--------------------- |------------:|----------:|----------:|------:|--------:|
| Regex                     | .NET 8.0             |  1,850.4 ns |  19.01 ns |  17.78 ns |  0.37 |    0.01 |
| StartsWithOrdinal         | .NET 8.0             |    467.5 ns |   6.62 ns |   6.19 ns |  0.09 |    0.00 |
| StartsWithInvariant       | .NET 8.0             |  1,582.4 ns |  27.14 ns |  24.06 ns |  0.32 |    0.01 |
| StartsWithCultureSpecific | .NET 8.0             |  2,395.2 ns |  21.98 ns |  20.56 ns |  0.48 |    0.01 |
| Handrolled                | .NET 8.0             |    409.7 ns |   7.86 ns |   7.72 ns |  0.08 |    0.00 |
| Regex                     | .NET Framework 4.7.2 |  7,186.8 ns | 133.12 ns | 124.52 ns |  1.43 |    0.04 |
| StartsWithOrdinal         | .NET Framework 4.7.2 |  4,960.0 ns |  97.68 ns | 127.01 ns |  1.00 |    0.00 |
| StartsWithInvariant       | .NET Framework 4.7.2 | 12,785.4 ns | 170.13 ns | 159.14 ns |  2.54 |    0.05 |
| StartsWithCultureSpecific | .NET Framework 4.7.2 | 16,590.5 ns | 193.12 ns | 180.65 ns |  3.30 |    0.09 |
| Handrolled                | .NET Framework 4.7.2 |  1,613.9 ns |  30.56 ns |  28.58 ns |  0.32 |    0.01 |

<details>

```c#
        [Benchmark]
        public bool Handrolled()
        {
            bool any_is_special = false;

            foreach (string propertyName in EnvVars)
            {
                any_is_special |= HandrollMatch(propertyName);
            }

            return any_is_special;
        }

        private bool HandrollMatch(string input)
        {
            ReadOnlySpan<char> slice = input.AsSpan();

            switch (slice[0])
            {
                case 'M' or 'm':
                    if ((uint)slice.Length < 7 ||
                        !slice.Slice(1).StartsWith("sbuild".AsSpan(), StringComparison.OrdinalIgnoreCase)) // Match the string "sbuild" (ordinal case-insensitive)
                    {
                        return false; // The input didn't match.
                    }

                    return true;

                case 'C' or 'c':
                    if ((uint)slice.Length < 8 ||
                        !slice.Slice(1).StartsWith("omplus_".AsSpan(), StringComparison.OrdinalIgnoreCase)) // Match the string "omplus_" (ordinal case-insensitive)
                    {
                        return false; // The input didn't match.
                    }

                    return true;


                case 'D' or 'd':
                    if ((uint)slice.Length < 7 ||
                        !slice.Slice(1).StartsWith("otnet_".AsSpan(), StringComparison.OrdinalIgnoreCase)) // Match the string "otnet_" (ordinal case-insensitive)
                    {
                        return false; // The input didn't match.
                    }

                    return true;

                default:
                    return false; // The input didn't match.
            }
        }
```

</details>

I'm pretty surprised by this result; I expected the generated/compiled regex to be basically equivalent to that. Maybe there's some missing optimization for `IsMatch` causing it to do a bunch of extra work?

In any case I'll go back to the three StartsWith calls; "handrolled" isn't worth the maintenance but might as well pick the best easy option since we all got sniped here :)

Reply:
WOW

I'm impressed by such a level of dedication to this code that's being called handful of times per build :-)

Reply:
We are all nerds and we know it 😛